[Feat] 투표 도메인 CRUD 기능 개발

IToriginal opened the pull request:
## 🤔 Motivation
- 투표 도메인 관련하여 타 도메인과 마찬가지의 구조로 CRUD 기능 개발을 완료하였습니다.

<br>

## 💡 Key Changes
### 1. 투표 생성 기능 개발 `POST` `/api/v1/vote`
[Required Data]
- `memberId`: 투표자 고유 식별자
- `boardId`: 게시글 고유 식별자
- `boardImage`: 투표할 이미지 고유 식별자

비즈니스 단에서 `createVote()`메서드를 통해 생성에 대한 로직을 수행하게 됩니다. 요구 데이터 값이 정상적이지 않다면 각 도메인에서 `NOT_FOUND` 예외를 발생하도록 구현하였습니다.

정상적인 데이터 값으로 간주되면, `VoteCreateResponse` 응답 DTO를 통해서 아래와 같이 응답 데이터를 결과로 반환 받습니다.
```
{
		"code": 200,
		"message": "성공적으로 투표를 완료하였습니다.",
		"data": {
			"voter": "woogi", // 투표자: 닉네임
			"boardId" : 1, // 게시글 고유 식별자
			"boardImageId" : 2 // 투표한 게시글 이미지 고유 식별자
		}
}
```

### 2. 투표 조회 기능 개발 `GET` `/api/v1/vote/{boardId}`
[Required Data]
- `memberId`: 회원 고유 식별자, `RequestHeader`로 `memberId`를 전달
- `boardId`: 조회할 게시글의 고유 식별자
투표를 진행한 회원만 아래의 결과를 받을 수 있는 검증 로직이 구현되어 있습니다. `validateVoter()`메서드를 통해서 투표를 진행한 회원인지를 검증하게 됩니다. 만약 투표를 진행하지 않았다면, 권한이 없는 회원으로 간주하도록 예외를 구현하였습니다.
```
{
		"code": "OK",
		"message": "성공적으로 게시글의 투표 상태를 조회하였습니다.",
		"data": {
				"boardId": 1,
				"totalVotes" : 2,
				"voteCounts" : {
						"image-1" : 1,
						"image-2" : 1,
						"image-3" : 0
				}
		}
}
```

### 3. 투표 수정 기능 개발 `PATCH` `/api/v1/vote/{voteId}`
[Required Data]
- `voteId`: 수정할 투표의 고유 식별자
- `memberId`: 회원 고유 식별자
- `BoardImage`: 게시글에 등록된 BoardImage 객체

자신이 등록한 투표가 맞는지 검증을 시도하고 검증 후에 수정할 BoardImage로 수정을 하도록 구현하였습니다.
```
{
		"code": 200,
		"message": "투표가 성공적으로 수정되었습니다.",
}
```

### 4. 투표 삭제 기능 개발 `DELETE` `/api/v1/vote/{voteId}`
[Required Data]
- `voteId`: 투표의 고유 식별자
- `memberId`: 회원 고유 식별자

수정 기능과 마찬가지로 자신의 투표인지 검증을 하고 검증 후에 투표 삭제(취소) 기능을 수행하도록 구현하였습니다.
```
{
		"code": "OK",
		"message": "성공적으로 투표를 삭제 하였습니다."
}
```

<br>

## 👨‍👨‍👧‍👦 To Reviewers
- @langoustinee @byeongJoo05
- 댓글 도메인을 개발한 구조와 대부분 동일하게 개발되었습니다. 
- 1차 기능 개발로 기본적인 기능을 수행하는데 초점이 맞춰져 있습니다. 투표 도메인도 이번 PR 이후로 리팩토링을 지속적으로 진행하여 개선되는 방향을 가져가면 좋겠습니다.
- 비즈니스 로직 개발 검증 차원으로 테스트코드가 몇 가지 추가되어 있는데 이는 기능 확인 차 작성된 코드이므로, 테스트코드는 다음 테스트코드 개발 이슈에서 변경 가능성이 있습니다.
- 추가적으로 개선할 수 있는 부분이나 주의해야 할 사항이 있다면 알려주시면 감사하겠습니다.

#129 
close: #133

IToriginal committed:
:sparkles: feat: 투표 생성 비즈니스 로직 개발

IToriginal committed:
:white_check_mark: test: 투표 생성 테스트 코드 추가

IToriginal committed:
:sparkles: feat: 투표 조회 비즈니스 로직 개발

IToriginal committed:
:white_check_mark: test: 투표 조회 테스트 코드 추가

IToriginal committed:
:sparkles: feat: 투표 도메인 예외처리

IToriginal committed:
:sparkles: feat: 투표 수정 비즈니스 로직 개발

IToriginal committed:
:white_check_mark: test: 투표 수정 테스트 코드 추가

IToriginal committed:
:sparkles: feat: 투표 삭제 비즈니스 로직 개발

IToriginal committed:
:white_check_mark: test: 투표 삭제 테스트 코드 추가

IToriginal commented on src/main/java/com/kakao/saramaracommunity/vote/controller/VoteController.java:
```diff
@@ -11,4 +28,88 @@
+     */
+    @GetMapping("/{boardId}")
+    public ResponseEntity<ApiResponse<VotesReadInBoardResponse>> getBoardVotes(
+            @RequestHeader("memberId") Long memberId,
```
`header`를 통해 `memberId`를 전달하도록 구현해보았습니다. 

앞서 `memberId`는 Dto를 통해 함께 전달하여 검증하는 방식을 적용했는데 조회 기능 개발에서는 `@RequestHeader`로 `memberId`를 전달하여 투표자를 확인하는 방식을 적용해보았습니다.

저는 지금까지 조회 기능에는 RequestData 클래스를 만들지 않았기 때문에 이와 같은 방식을 적용해보았는데 어떤 방법이 좋을지 의견주시면 감사하겠습니다.

langoustinee replied:
저는 `@RequestHeader`에 사용자 정보를 포함하는 것은 굳이 필요없다고 생각되네요.

일반적으로 HTTP 헤더는 메타데이터를 전송하는데 사용되는 것으로 알고 있어서, 데이터를 위한 데이터를 보내는 데 사용하는 것이 일반적인 것 같아요. 가령 콘텐츠 유형(Content-Type), 인증 토큰(Authorization Token) 등을 헤더로 보내는 것이 더 헤더답다고 해야할까요?

그리고 프론트엔드와의 교착점도 생길테구요! 토큰과 같은 귀중한 정보가 아닌 이상 웬만하면 파라미터를 DTO나 필드로 보내보는 것은 어떨까요?

IToriginal replied:
그렇군요! 의견 감사합니다. 저도 앞서 DTO의 방식을 선호했는데, `@RequestHeader`라는 것이 존재하여 사용해보았습니다. 

`@RequestHeader`를 사용한 경우를 좀 더 말씀 드려보자면, 아래와 같이 생각해 간단한 요구사항인 `조회API`에 사용하기 적합하다 생각했습니다.
- 별도의 객체를 정의하거나 매핑할 필요가 없어서 간단하게 사용할 수 있다. (=간단한 API이다.)
- 여러 개를 사용하면 가독성이 떨어지는데 조회 API의 경우 memberId 한 필드만 필요로 한다.
- 헤더에서 직접 값을 추출하기 때문에 헤더에 값이 변경되거나 추가되는 경우 코드를 수정해야 하는 유연성이 떨어지는 단점이 존재하지만, memberId는 고유 식별자로 변경될 가능성이 없고 조회에 추가될 데이터도 없기 때문에 단점이라 볼 수 없다.

이처럼 제가 생각했을 때는 조회에 대한 API에서는 읽기에 대한 기능으로서 `@RequestHeader`를 사용해서 데이터를 전달하는 방법이 DTO를 거치지 않기 때문에 좀 더 효율적이라 생각하게 되었습니다.

말씀해주신 것 처럼 HTTP헤더의 경우 메타데이터를 전송하는데 제가 `memberId`를 전달하게 된 이유는 아래와 같습니다.
- `memberId`를 전달하는 이유: 투표 조회에 투표를 한 회원인지를 판별하기 위함

이처럼 메타데이터로서의 역할을 하지 않나? 생각을 하고 있어요.. 
추가로 토큰과 같은 것이 좀 더 헤더 답지 않을까하셨는데, 이는 회원의 대한 인증 방법에 따라 토큰을 통한 권한 확인 기능이 존재한다면 이를 사용할 듯 합니다.. (memberId의 필요 여부이기 때문에..)

한가지 궁금한 점은 프론트엔드와의 교착점이 생긴다 하셨는데, 제가 이에대해 정보를 찾는데 어려움이 있네요.. 이에 대해 좀 더 생각할 수 있도록 알려주실 수 있을까요? (죄송하게도 프론트쪽은 어떤 상황인지 떠오르지가 않습니다.🥲)

langoustinee replied:
프론트엔드와의 교착점이라고 전달드렸던 것은 프론트엔드 개발자와의 협의도 필요한 부분이고, 다른 API에서는 `memberId`를 DTO로 보내는 것이 대부분인데, 투표 조회 API에서만 Header로 보내는 것이 혼동의 여지가 있어 보여서요! 

말씀해주신대로 `memberId`를 전달하는 이유가 투표 조회에 투표를 한 회원인지를 판별하기 위함이라면, 다른 API에도 같은 방식을 취해야 하지 않을까? 라는 생각이 드네요. 어떻게 생각하시나요? 다른 API 중에서도 memberId만을 DTO로 보내는 API들이 있긴 하거든요.

제가 생각하기엔 1차적으로 정한 API 설계와 수정 작업을 추가적으로 해주어야 하기 때문에 현시점에서 MVP 개발을 위한 우선순위로는 조금 미뤄두는 게 어떨까싶어요. 저는 현시점에서 프로토타입을 우선적으로 도출해보고 싶은 마음이 큽니다 ㅠ

IToriginal replied:
개발 커뮤니티 채널에 이와 관련해 논의를 해보기 위해 아래와 같이 채널에 남기게 되었습니다!
이후 논의된 결과로 진행해보겠습니다!

<img width="1041" alt="image" src="https://github.com/four-uncles/saramara-community-server/assets/117193889/4ba07ee7-23f3-4224-a7c5-8729cfb4acb5">

IToriginal commented on src/main/java/com/kakao/saramaracommunity/vote/dto/business/response/VotesReadInBoardResponse.java:
```diff
@@ -0,0 +1,19 @@
+public record VotesReadInBoardResponse(
+        Long boardId,
+        Long totalVotes,
+        Map<String, Long> voteCounts // 이미지 경로, 투표수
```
Map을 활용하여 KEY(이미지 경로, ex. image-1) - VALUE(투표 수, ex. 4)로 응답하도록 타입을 지정하였습니다.

langoustinee replied:
`totalVotes`와 `voteCounts`를 나누어서 응답할 필요가 있을까요? 

서버에서 총 투표수를 구해서 클라이언트로 내려줄지, `voteCounts`만으로 클라이언트에서 총 투표수를 계산하여 사용할지 의사결정이 필요하겠네요.

IToriginal replied:
`totalVotes`의 필요여부는 사실 서버와 클라이언트 어디서 해결할지의 문제인 것 같습니다. 또한, 메모리 비용에도 영향이 있겠죠.

하지만, 인덱스 1개에 크게 차이는 나지 않을 것으로 판단하고, 클라이언트에서 JSON 데이터를 다시 받아 통계하는 것보다, 서버에서 voteCounts를 통계하면서 total 통계를 내는 것이 좀 더 개발 비용에서 아낄 수 있지 않나 생각합니다.

이에 대해서 의견을 여쭙고, 결정에 따라 필드를 삭제할지 결정해보겠습니다! 클라이언트 개발자의 의견도 궁금하네요?!

langoustinee replied:
넵. 해당 부분에 대한 것은 프론트엔드 개발자와 논의해본 후 의사결정된 대로 진행하겠습니다.

IToriginal commented on src/main/java/com/kakao/saramaracommunity/vote/repository/VoteRepository.java:
```diff
@@ -1,8 +1,26 @@
+     * @param boardId 게시글 고유 식별자
+     * @return 이미지 경로 - 투표 수의 객체 목록
+     */
+    @Query("SELECT i.path, COALESCE(COUNT(v), 0) " +
+            "FROM BoardImage i " +
+            "LEFT JOIN Vote v ON i.id = v.boardImage.id AND v.board.id = :boardId " +
+            "GROUP BY i.path")
+    List<Object[]> getVotesByBoard(@Param("boardId") Long boardId);
```
게시글 이미지(BoardImage i) 테이블의 `path` 필드와 투표(Vote v) 테이블의 갯수를 정렬하도록 하였고, `COALESCE` 함수를 사용하여 테이블이 없다면 0을 조회하도록 하였습니다.

`LEFT JOIN` 함수를 통해 `BoardImage` 테이블을 기준으로 `Vote` 테이블과 연결하였고, `GROUP BY` 함수를 사용하여  게시글 이미지의 `path`를 기준으로 결과를 그룹화하도록 쿼리를 작성하였습니다.

쿼리 실행 결과 예:
```
+--------+-----------------------+
|  path  | COALESCE(COUNT(v), 0) |
+--------+-----------------------+
|image-1 | 2                     |
|image-2 | 1                     |
|image-3 | 3                     |
+--------+-----------------------+
```

메소드 실행 결과 예:
Object[]을 가지는 List 타입으로 인덱스[0]의 path 인덱스[1]의 투표 수(COALESCE(COUNT(v), 0)를 담아 리스트로 반환합니다.

langoustinee replied:
작성하신 쿼리메서드 잘 읽어보았어요. 해당 메소드로 구하고자 하는 데이터는 **특정 게시글의 각 이미지에 대한 투표 수**라고 이해했어요. 아래 2가지 정도 피드백을 드려볼 수 있을 것 같네요.

### 첫 번째 리뷰
먼저 해당 쿼리의 From절을 보면 `BoardImage`를 주 조회 대상으로 지정하신 것으로 보이는데, `VoteRepository`인데, `BoardImage`라는 다른 엔티티를 기준으로 작성한다면 `BoardImageRepository`의 쿼리메서드로 개발하는 것과 헷갈릴 여지가 있을 것 같아요. (물론 성능상 어떤 이점이 차이나는지 아직 제대로 파악은 안된 상태라, 정답은 없습니다. 큐큐) 

그래서 주 조회대상을 `BoardImage`가 아닌 `Vote`에서 조회를 하되 `BoardImage`를 Join으로 가져오는 것이 `VoteRepository`의 개발될 쿼리메서드의 원칙이 아닌가 싶습니다.

### 두 번째 리뷰
사실 기존 게시글 등록시 클라이언트로 내려줄 때 게시글 이미지와 관련된 정보는 아래와 같이 이미지 객체 주소(String)만을 내려주어요. 

```json
{
		"code": 200,
		"message": "성공적으로 게시글 정보를 조회하였습니다.",
		"data": {
			"boardId": 1L,
			"title": "title ...",
			"content": "content is ...",
			"categoryBoard": "VOTE",
			"writerNickname": "lango",
			"writerEmail": "lango@test.com",
			"deadLine": "2023-10-13T10:12:00",
			"images": [
									"imageURL-1",
									"imageURL-2",
									"imageURL-3",
									"imageURL-4",
									"imageURL-5"
								]
		}
}
```

그런데, 이때 이미지 경로뿐만 아니라, 이미지 아이디도 함께 내려주도록 고쳐볼게요.

```json
{
		"code": 200,
		"message": "성공적으로 게시글 정보를 조회하였습니다.",
		"data": {
			"boardId": 1L,
			"title": "title ...",
			"content": "content is ...",
			"categoryBoard": "VOTE",
			"writerNickname": "lango",
			"writerEmail": "lango@test.com",
			"deadLine": "2023-10-13T10:12:00",
			"images": [
									{"3": "imageURL-1"},
									{"4": "imageURL-2"},
									{"5": "imageURL-3"},
									{"6": "imageURL-4"},
									{"7": "imageURL-5"}
								]
		}
}
```

위와 같이 응답하게 된다면 게시글 이미지 ID를 참조할 수 있어서 개발하신 투표 조회 API에서는 굳이 BoardImage를 Join하지 않고도 쿼리를 작성해볼 수 있을 것 같네요.

```java
@Query("SELECT v.boardImage.id, COUNT(v) FROM Vote v WHERE v.board = :board GROUP BY v.boardImage")
```

요런식으로요. 

> 위 쿼리 내용은 수도코드로 작성한 것이니 참고만 해주세요.

그래서 정리하자면, **게시글의 이미지 정보를 클라이언트로 전달할 때 이미지 경로만 줄 것이냐, 게시글이미지 ID 정보도 줄 것이냐가 관건**이겠네요. 한번 이야기 해봅시다. 프론트엔트 개발자와도 논의해보아야 할 것 같구요!

IToriginal replied:
정성스러운 리뷰 감사합니다. 순서대로 답변 드리도록 하겠습니다.

### 첫번째 리뷰에 대한 답변: 
말씀처럼 Vote가 주가 되는 도메인의 쿼리문이기 때문에 FROM에서 BoardImage가 오해의 소지가 있겠다는데 공감했습니다. 때문에 아래와 같이 기준이 되는 테이블을 투표 테이블인 `Vote v`로 수정하고, 기준이 변경됨에 따라 `LEFT JOIN` -> `RIGHT JOIN`으로 수정하겠습니다. 예상되는 코드는 아래와 같으므로 참고 해주시면 감사하겠습니다.
```java
@Query("SELECT i.path, COALESCE(COUNT(v), 0) " +
            "FROM Vote v " +
            "RIGHT JOIN v.boardImage i ON v.board.id = :boardId " +
            "GROUP BY i.path")
List<Object[]> getVotesByBoard(@Param("boardId") Long boardId);
```

### 두번째 리뷰에 대한 답변:
이미지에 대한 고유 아이디를 함께 전달해줄 필요는 없을 것 같아요. 이유는 게시글에는 어차피 BoardImage에 대한 객체 자체를 가지고 있기 때문에 해당 객체에서 투표 도메인이 필요로 한다면 해당 id를 꺼내서 사용하면 될 것이라 생각합니다. 

다만, 제가 `boardImage`의 `id`가 아닌 `path`를 기준으로 API를 설계 했던 것은 `테스트 코드`가 좀 더 유연하게 작성될 수 있지 않을까 하는 생각이였어요.
아시다시피 저희는 `고유 식별자`를 왠만하면 노출하지 않으려하고, 테스트 코드에서도 `id`에 대한 어려움이 생각보다 빈번하게 발생하기 때문이 였습니다.
저희 컨셉은 AWS S3 Bucket을 이용한 `path` 이므로 이를 활용해도 된다고 생각했어요. (이미지 경로에 대한 유일성이 보장될 것이라는 생각)

이에 대해서는 한번 논의를 거쳐 `id`를 사용하여 `Map`을 형성하는게 좋겠다는 판단이 되면, 해당 기능을 id로 mapping할 수 있도록 수정해보겠습니다!

langoustinee commented:
@IToriginal PR 내용을 보기전 PR 형식과 관련하여 간단한 질문을 드리고 싶어요!

- 이전 스프린트에서 결정했던 내용으로는 이슈는 `[도메인/작업범위]` 프리픽스로, PR에는 `[작업범위]` 프리픽스만으로 타이틀 내용을 작성하기로 했던 것 같은데, 혹시나 제가 잘못 이해하고 진행했을 까요? 이전 댓글 관련 PR도 그렇게 작성하셨길래 여쭤봅니다.
- 그리고 Key Changes 내용에 기입하신 API 명세에 관련된 내용들은 노션에 이미 기입된 내용이라면 링크만 두셔도 무방할 것 같아요.

작성해주신 PR 리뷰해보고 Request Changes 남기겠습니다!!

IToriginal commented on src/main/java/com/kakao/saramaracommunity/vote/service/VoteServiceImpl.java:
```diff
@@ -0,0 +1,158 @@
+        return totalVotes;
+    }
+
+    private void validateVoter(Long memberId, Long boardId) {
+        Optional<Vote> existVote = voteRepository.findByMemberIdAndBoardId(memberId, boardId);
+        if (existVote.isPresent()) {
+            log.info("[VoteServiceImpl.class] 해당 게시글의 투표자 검증이 완료되었습니다.");
+        } else {
+            log.info("[VoteServiceImpl.class] 투표 상태를 조회할 수 있는 상태가 아닙니다. 투표를 완료해주세요.");
+            throw new MemberBusinessException(UNAUTHORIZED_TO_MEMBER);
+        }
+    }
```
투표 조회시에 해당 게시글에 대한 투표를 진행했을 경우와 투표를 하지 않았을 경우, 조회에 대한 기능을 검증하는 로직인데, 해당 부분이 필요한지 의문이 들었습니다.

우선은 백엔드 개발로서 구현을 해두긴 했는데, Front 에서 처리할 수 있을 것이라는 생각이 들었습니다. 또한, 권한에 대한 기능으로 생각해서 이를 투표 도메인에서 처리하는 것이 맞을지 의문이 였습니다.

결론은 백엔드에서 1차 구현은 했지만, 해당 API를 권한으로서 관리하여 권한에 걸리지 않는 사용자만이 해당 API를 프론트에서 호출해주면 되지 않을까? 하는 생각이 들었습니다.

langoustinee replied:
일단 서버에서 검증할 수 있을만한 것들은 최대한 해보시죠! 저희의 Grow Up을 위해서요!

`validateVoter` 내부의 조건문은 다음과 같이 고쳐보면 더 명시적으로 개선될 것 같아요.

```java
private void validateVoter(Long memberId, Long boardId) {
    Optional<Vote> existVote = voteRepository.findByMemberIdAndBoardId(memberId, boardId);
    if (!existVote.isPresent()) {
        log.info("[VoteServiceImpl.class] 투표 상태를 조회할 수 있는 상태가 아닙니다. 투표를 완료해주세요.");
        throw new MemberBusinessException(UNAUTHORIZED_TO_MEMBER);
    }
    log.info("[VoteServiceImpl.class] 해당 게시글의 투표자 검증이 완료되었습니다.");
```

위와 같이요. 물론 최하단 로그는 해당 메서드 호출 이후로 이동시킬 수도 있겠네요.

IToriginal replied:
투표자 검증에 대한 의견 감사합니다!

사실 `isPresent()`의 부정적 응답으로 `isEmpty()`를 사용할 수가 있어서 말씀해주신 것과 같이 작성해봤었는데 가독성을 해치는 느낌을 받아 `if - else`를 사용했던 것이 있습니다. 
하지만, 리뷰해주시는 것을 보니 제 착각이였던 것 같네요. 아래와 같이 수정해보겠습니다. 참고 부탁합니다!
```java
private void validateVoter(Long memberId, Long boardId) {
    Optional<Vote> existVote = voteRepository.findByMemberIdAndBoardId(memberId, boardId);
    if (existVote.isEmpty()) { // = !existVote.isPresent() 
        log.info("[VoteServiceImpl.class] 투표 상태를 조회할 수 있는 상태가 아닙니다. 투표를 완료해주세요.");
        throw new MemberBusinessException(UNAUTHORIZED_TO_MEMBER);
    }
    log.info("[VoteServiceImpl.class] 해당 게시글의 투표자 검증이 완료되었습니다.");
}
```

IToriginal commented:
> @IToriginal PR 내용을 보기전 PR 형식과 관련하여 간단한 질문을 드리고 싶어요!
> 
> * 이전 스프린트에서 결정했던 내용으로는 이슈는 `[도메인/작업범위]` 프리픽스로, PR에는 `[작업범위]` 프리픽스만으로 타이틀 내용을 작성하기로 했던 것 같은데, 혹시나 제가 잘못 이해하고 진행했을 까요? 이전 댓글 관련 PR도 그렇게 작성하셨길래 여쭤봅니다.
> * 그리고 Key Changes 내용에 기입하신 API 명세에 관련된 내용들은 노션에 이미 기입된 내용이라면 링크만 두셔도 무방할 것 같아요.

@langoustinee 
1. 앗! 아닙니다. 제가 잊고 있었던 것 같아요! 태준님이 잘 사용하고 계셨습니다. 습관처럼 도메인을 작성하게 됐네요.. 말씀해주셔서 다시 기억이 났습니다. 해당 부분은 수정하고 이후 작업부터 체크해보겠습니다!
2. 넵 확인하였습니다!👍🏻

langoustinee commented on src/main/java/com/kakao/saramaracommunity/vote/entity/Vote.java:
```diff
@@ -30,19 +33,43 @@ public class Vote extends BaseTimeEntity {
-    @Builder
-    public Vote(Member member, BoardImage boardImage, Board board) {
+    @Builder(access = AccessLevel.PRIVATE)
+    public Vote(Member member, Board board, BoardImage boardImage) {
```
생성자의 접근제한자는 public인데 Builder의 접근제한자를 private으로 설정하신 것을 보았어요. 아래 정적 팩토리 메서드를 사용하기 위해 의도하신 것 같은데, 생성자의 접근제한자가 public일 필요가 있을까요?

정잭 팩토리 메서드인 of() 메서드를 활용한 객체 생성 전략을 취하는 의도시라면 생성자의 접근제한자는 private으로 설정하는 것이 어떨까요?

IToriginal replied:
헙 발견해주셔서 감사합니다! 휴먼 이슈네요.. 🥲

langoustinee commented on src/main/java/com/kakao/saramaracommunity/vote/service/VoteServiceImpl.java:
```diff
@@ -0,0 +1,158 @@
+
+    @Override
+    public VoteCreateResponse createVote(VoteCreateServiceRequest request) {
+        log.info("[VoteServiceImpl.class] 요청에 따라 투표를 시도합니다.");
```
아주 사소한 내용이긴 하지만..
로그에서 출력할 내용 프리픽스에서 어떤 서비스인지만 알면 될 것 같아서 `.class`라는 확장자명은 제외해도 무방할 것 같아요.

langoustinee commented on src/main/java/com/kakao/saramaracommunity/vote/dto/business/response/VotesReadInBoardResponse.java:
```diff
@@ -0,0 +1,19 @@
+
+import java.util.Map;
+
+public record VotesReadInBoardResponse(
```
`VotesReadInBoardResponse` 이라는 네미잉을 기획하신 의도를 알고 싶어요!
`In`의 의미가 무엇인가요?

IToriginal replied:
투표 테이블의 전체를 조회하는 것이 아니라 
한 게시글 **안**에 투표기 때문에 `In`을 명시해줬습니다! ("게시글 안의 투표를 조회한다")

langoustinee replied:
오호, 그렇군요. 
저의 지극히 개인적인 생각으로는 저희 프로젝트에서 투표를 조회하는 개념이 현재 기획 상으로는 게시글 상세화면와 메인 화면일텐데, 그 중에서도 특정 게시글의 투표를 조회하는 것은 게시글 상세에서만 사용할 것 같아요. 

제가 이해하고 있는 맥락으로는 그렇다보니 `VotesReadInBoardResponse`라는 이름도 좋지만, 특정 게시글 조회에 대한 DTO 명으로는 `VotesResponse`, 메인화면에 필요한 여러 투표수의 경우는 `List<VotesResponse>` 등으로 단일 DTO를 목록으로 담아서 재사용하는 방식으로도 사용해볼 수 있겠다. 생각이 드네요. 물론 `VotesResponse`를 목록으로 가공할 수 있을지는 추가적인 기획과 검토가 필요하겠지만요.

일단은 당장 중요한 문제는 아니니 추후 이야기해보시죠!

langoustinee commented on src/main/java/com/kakao/saramaracommunity/vote/service/VoteServiceImpl.java:
```diff
@@ -0,0 +1,158 @@
+                .orElseThrow(() -> new BoardBusinessException(BOARD_IMAGE_NOT_FOUND));
+    }
+
+    private Map<String, Long> getVoteCounts(List<Object[]> votes) {
+        Map<String, Long> voteCounts = new LinkedHashMap<>();
+
+        for (Object[] vote : votes) {
+            String imagePath = (String) vote[0]; // 이미지 경로
+            Long count = (Long) vote[1]; // 투표 수
+            voteCounts.put(imagePath, count);
+        }
+        return voteCounts;
+    }
```
`getVoteCounts`에서 `LinkedHashMap`을 사용하신 이유를 간단하게 알고 싶어요.

IToriginal replied:
일반적인 HashMap을 사용하게 되면, 데이터가 이미지의 순서를 보장해주지 않습니다. 게시글을 보면 이미지의 순서대로 이미지객체를 생성해주는 것으로 이해하고 있습니다. 이와 동일한 결과를 응답하기 위해서는 순서가 보장되는 것이 좋겠다고 생각했습니다.

HashMap 결과 예시
```
{
    "image-2": 1,
    "image-1": 0
}
```

LinkedHashMap 결과 예시:
```
{
    "image-1": 0,
    "image-2": 1
}
```

물론, 클라이언트에서 어느 기준을 통해서 화면에 출력할 것이기 때문에 불필요할 수 있지만, 순서가 보장되는 것이 서버단에서도 데이터를 핸들링하는데 편리할 것이라는 생각 합니다. (순서가 보장되는 것이 테스트 코드에서도 용이할 것)

langoustinee replied:
엇 뭔가 비즈니스 맥락 차원에서 짚고 넘어가야 할 부분이 있어요!

> 게시글을 보면 이미지의 순서대로 이미지객체를 생성해주는 것으로 이해하고 있습니다.

이전 스프린트에서 게시글 이미지의 순서를 고려하는 것은 아예 제외하고 비즈니스를 개발하자고 결정되어 **게시글 이미지의 순서는 전혀 고려하지 않았을뿐더러 고려할 필요**가 없어요.

그냥 게시글 내에는 어떤 이미지, 어떤 이미지, 어떤 이미지를 무작위로 뿌리는 방향으로 나아가기로 이야기했었는데 부디 기억나셨으면 좋겠습니다 😅

그래서 이미지 순서 보장해줄 필요없이 무작위로 담아주셔도 무방합니다!!!

```json
{ 
    "이런 이미지 경로": "투표 수",
    "저런 이미지 경로": "투표 수",
    "그냥 이미지 경로": "투표 수",
} 
```

IToriginal replied:
맞습니다. 물론 해당 내용을 인지하고 있습니다.👍🏻

Board 객체에서 BoardImage 객체를 생성할 때의 순서(생성순)에 맞게 List가 생성되는 것을 볼 수 있었습니다. (ArrayList)
이와 관련하여 **개발의 생산성을 높이기 위해** `LinkedHashMap`을 사용하게 되었습니다.

LinkedHashMap을 사용하면 HashMap보다 메모리 측면에서 비효율적인데,
제가 생각하기로는 Board 객체에 저장된 BoardImage의 순서와 투표 통계의 순서가 일치하는 것이 좀 더 개발하기 편리할 것으로 생각하게 되었습니다.

때문에 생산성에 큰 영향을 주지 못하거나 메모리로 인한 성능 저하가 발견된다면 이를 수정하는 것이 좋을 듯합니다.👍🏻

langoustinee reviewed (changes requested):
투표 관련 비즈니스 개발사항 잘 읽었습니다. 고민하고 코딩했다는 것이 보여 기쁩니다.
달아주신 코멘트 모두 확인하고 제 의견을 살포시 달아두었어요. 확인 바래요!

추가로 테스트 케이스의 경우 직접 머지받고 테스트 수행 결과를 보며 확인하는 것이 나을 것으로 보여져서 일단 비즈니스 로직에 초점을 두고 리뷰를 해보았음을 참고해주시길 바라요.

IToriginal committed:
:recycle: Refactor: Vote 객체 생성자 접근 제어자 변경

- 코드 리뷰 반영

IToriginal committed:
:fire: Remove: 불필요한 로그 메시지 일부 삭제

- 코드 리뷰 반영

langoustinee commented on src/main/java/com/kakao/saramaracommunity/vote/controller/VoteController.java:
```diff
@@ -11,4 +28,88 @@
 @RequestMapping("/api/v1/vote")
```
PR 코멘트 숨김 처리 테스트입니다.

langoustinee reviewed (dismissed):
금일 스프린트에서 이야기한 내용을 기준으로 요구사항이 충족되었다고 판단되어서 병합 승인 하겠습니다~

IToriginal committed:
:recycle: Refactor: 투표 조회시 투표자 검증 로직 간결화

- 코드 리뷰 사항 반영

IToriginal committed:
:recycle: Refactor: JPQL 쿼리 기준 도메인 수정

- 코드 리뷰 사항 반영

IToriginal committed:
:fire: Remove: 회원 고유 식별자 헤더 정보 삭제

- 코드 리뷰 반영
- 투표자(사용자) 권한 검증을 위한 보안성 강화 관련 개발 예정 (memberId 고유 식별자 활용 안)
- 투표 조회 테스트 코드는 해당 검증 로직이 추가된 후 개발 예정

IToriginal committed:
:recycle: Refactor: 투표수 통계 기능에 사용되는 메서드 HashMap 자료구조 리팩토링

- 코드 리뷰 사항 반영
- 이미지는 순서에 무관하므로 LinkedHashMap 사용이 불필요하다고 판단하였음
